[CAS-1066] Message list component doc (#1997)
GetStream/stream-chat-android: docusaurus/docs/Android/03-ui/04-components/03-message-list.md

1. Adding component to your xml layout hierarchy
2. Bind out-of-box `MessageListViewModel` and `MessageLisView`.

### Adding to xml layout
Adding `MessageListView` to your layout is easy as inserting following lines to your layout hierarchy (example for `ConstraintLayout`):

```xml
<io.getstream.chat.android.ui.message.list.MessageListView
        android:id="@+id/message_list_view"
        android:layout_width="0dp"
        android:layout_height="0dp"
        app:layout_constraintBottom_toBottomOf="parent"
        app:layout_constraintLeft_toLeftOf="parent"
        app:layout_constraintRight_toRightOf="parent"
        app:layout_constraintTop_toTopOf="parent"
        />
``` 

### ViewModel and binding
UI components provide out-of-box a view model for `MessageListView` and the `bindView` extension function that makes default setup:

```kotlin
class MessageListFragment : Fragment() {

    private lateinit var messageListView: MessageListView

    fun bindView() {
        val channelCID = "messaging:123"
        val viewModel = MessageListViewModel(cid = channelCID)

>         viewModel.bindView(messageListView, this)

Reviewer: We should showcase using the factory instead of creating a ViewModel directly, otherwise clients are gonna copy this code and not have their ViewModels retained. We can probably also drop the Fragment context of this code snippet (especially since we're not initializing the `lateinit var` anyway) - see what the Channel List docs do for both of these.

| --- | --- |
|![Message_options_in light mode](../../assets/message_options_light.png)|![Message options in dark mode](../../assets/message_options_dark.png)|

> Most of the actions work out of the box but you can change their behavior using different listeners and handlers:

Reviewer 2: I cannot comment above this line but I would replace this sentence ```MessageListView comes with a set of actions out of the box that are available by long-pressing a message. There are multiple available actions:``` with ```MessageListView comes with a set of actions available out of the box by long-pressing a message:```. Moreover some of the listed actions are hidden by default and we also allow to Unmute users

> If you're not going to use out of the box `MessageListViewModel` with it's handlers or just want to override action handlers you should define these handlers:

Reviewer 2: ```suggestion
Default action handlers are set after binding ViewModel with the View. You can customize default behavior by overriding each of the following handlers:
```

Reviewer 2: and maybe before that - what do you think about explaining what handlers are?

Author: I think in the code sample there are nice comments which explain

            // Handle when some giphyAction is going to be performed
        }
        messageListView.setMessageRetryHandler { message: Message ->
            // Handle when some failed message is going to be retried
        }
        messageListView.setMessageReactionHandler { message: Message, reactionType: String ->
            // Handle when some reaction for message is going to be send
        }
        messageListView.setUserMuteHandler { user: User ->
            // Handle when a user is going to be muted
        }
        messageListView.setUserUnmuteHandler { user: User ->
            // Handle when a user is going to be unmuted
        }
        messageListView.setUserBlockHandler { user: User, cid: String ->
            // Handle when a user is going to be blocked in the channel with cid
        }
        messageListView.setMessageReplyHandler { cid: String, message: Message ->
            // Handle when message is going to be replied in the channel with cid
        }
        messageListView.setAttachmentDownloadHandler { attachment: Attachment ->
            // Handle when attachment is going to be downloaded
        }
    }
``` 

> ---

Reviewer 2: ```suggestion
```

> **NOTE**

Reviewer 2: ```suggestion
:::note
```

Author: Are u sure that it works? My preview doesn't show it correctly

Handlers must be set before passing any data to `MessageListView`. So if you don't use default binding, please, make sure you defined them.

> ___

Reviewer 2: ```suggestion
:::
```

### Listeners

> Except of required handlers you're able to set listeners to get events when something happens:

Reviewer 2: It might be worth it writing that the same as with Handlers - default ones are set when binding view model with view

```kotlin
fun setListeners() {
        messageListView.setMessageClickListener { message: Message ->
            // Listen to click on message events
        }
        messageListView.setEnterThreadListener { message: Message ->
            // Listen to events when enter thread associated with a message
        }
        messageListView.setAttachmentDownloadClickListener { attachment: Attachment ->
            // Listen to events when download click for an attachment happens
        }
    }
```

> Other available listeners for `MessageListView` can be found [here](https://github.com/GetStream/stream-chat-android/blob/5084b1528f15530782648de559d58de6d55045d5/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/MessageListListenerContainer.kt)

Reviewer 2: I think we can point to Dokka instead

> The full lists of available listeners and handlers are available [here (MessageListView)](https://getstream.github.io/stream-chat-android/stream-chat-android-ui-components/stream-chat-android-ui-components/io.getstream.chat.android.ui.message.list/-message-list-view/index.html).

Reviewer 2: This is probably no longer needed

## Customizations

If you want to setup appearance of this component by your design requirements you're free to do it.
There are two ways to change the style: using XML attributes and runtime changes.

### Customization with XML Attributes

> `MessageListView` provides a quite big set of xml attributes available for customization. The full list of them is available [here](https://github.com/GetStream/stream-chat-android/blob/main/stream-chat-android-ui-components/src/main/res/values/attrs_message_list_view.xml)

Reviewer 2: ```suggestion
`MessageListView` provides a quite big set of xml attributes available for customization. The full list of them is available [here](https://github.com/GetStream/stream-chat-android/blob/main/stream-chat-android-ui-components/src/main/res/values/attrs_message_list_view.xml).
```

    )
}
```

> ___

Reviewer 2: We could use:
:::note
:::

}
```

## Channel features flags

Some xml attributes provide possibility to enable/disable some features. For example:

> 1. R.attrs.streamUiReplyEnabled - defines if users can reply to messages

Reviewer 2: NIT: I would use unordered list

2. R.attrs.streamUiCopyMessageActionEnabled - defines if users can copy messages
3. R.attrs.streamUiEditMessageEnabled - defines if users can edit their messages
4. R.attrs.streamUiMuteUserEnabled - defines if users can mute others
5. R.attrs.streamUiDeleteMessageEnabled - defines if users can delete their messages

> 6. Others you can find the full list of attributes [here](https://github.com/GetStream/stream-chat-android/blob/main/stream-chat-android-ui-components/src/main/res/values/attrs_message_list_view.xml)

Reviewer 2: Do we have other flags? Maybe in that case it's worth it to enumerate all of them?

These attributes let you enable/disable configuration for channel features. E.g. if a channel supports message replies, but you disabled it via xml attributes then members of this channel won't see such option.
`MessageListView` provides you possibility to enable/disable these channel features in runtime.

```kotlin
fun disableChannelFeatures() {
        messageListView.setRepliesEnabled(false)
        messageListView.setDeleteMessageEnabled(false)
        messageListView.setEditMessageEnabled(false)
}
```
| Before | After |
| --- | --- |
|![message list options before](../../assets/message_list_options_before.png)|![message list options after](../../assets/message_list_options_after.png)|

## Message list item factory

> `MessageListItem` provides API for creating custom view holders. You can set custom view holder factory to `MessageListView`. Just extend `MessageListItemViewHolderFactory`, write your logic, make new instance and set to `MessageListView`.

Reviewer 2: ```suggestion
`MessageListItem` provides API for creating custom view holders. All you need to do is:
1. Extend  `MessageListItemViewHolderFactory`
2. Write your own logic for creating ViewHolders 
3. Make a new factory instance and set it to `MessageListView`.
```

> Let's consider an example when we want to create custom view holders for messages from other users that came less than 24 hours ago.

Reviewer 2: ```suggestion
Let's consider an example when we want to create custom view holders for messages that came from other users less than 24 hours ago.

```
